src/coord/src/transform_ast.rs

```diff
@@ -0,0 +1,356 @@
+    let from_ident = Ident::new(from.item.clone());
+    let to_ident = Ident::new(to.item.clone());
+    let qual_depth =
+        QueryIdentAgg::determine_qual_depth(&from_ident, Some(to_ident.clone()), query)?;
+    CreateSqlRewriter::rewrite_query_with_qual_depth(from, to, qual_depth, query);
+    // Ensure that our rewrite didn't didn't introduce ambiguous
+    // references to `to_name`.
+    match QueryIdentAgg::determine_qual_depth(&to_ident, None, query) {
+        Ok(_) => Ok(()),
+        Err(e) => Err(e),
+    }
+}
+
+fn ambiguous_err(n: &Ident, t: &str) -> String {
+    format!("{} potentially used ambiguously as item and {}", n, t)
+}
+
+/// Visits a [`Query`], assessing catalog item [`Ident`]s' use of a specified `Ident`.
+struct QueryIdentAgg<'a> {
+    /// The name whose usage you want to assess.
+    name: &'a Ident,
+    /// Tracks all second-level qualifiers used on `name` in a `HashMap`, as
+    /// well as any third-level qualifiers used on those second-level qualifiers
+    /// in a `HashSet`.
+    qualifiers: HashMap<Ident, HashSet<Ident>>,
+    /// Tracks all potential item names we encounter.
+    items: HashSet<Ident>,
+    /// Tracks the least qualified instance of `name` seen.
+    min_qual_depth: usize,
+    /// Provides an option to fail the visit if encounters a specified `Ident`.
+    fail_on: Option<Ident>,
+    err: Option<String>,
+}
+
+impl<'a> QueryIdentAgg<'a> {
+    /// Determines the depth of qualification needed to unambiguously reference
+    /// catalog items in a [`Query`].
+    ///
+    /// Includes an option to fail if a given `Ident` is encountered.
+    ///
+    /// `Result`s of `Ok(usize)` indicate that `name` can be unambiguously
+    /// referred to with `usize` parts, e.g. 2 requires schema and item name
+    /// qualification.
+    ///
+    /// `Result`s of `Err` indicate that we cannot unambiguously reference
+    /// `name` or encountered `fail_on`, if it's provided.
+    fn determine_qual_depth(
+        name: &Ident,
+        fail_on: Option<Ident>,
+        query: &Query,
+    ) -> Result<usize, String> {
+        let mut v = QueryIdentAgg {
+            qualifiers: HashMap::new(),
+            items: HashSet::new(),
+            min_qual_depth: usize::MAX,
+            err: None,
+            name,
+            fail_on,
+        };
+
+        // Aggregate identities in `v`.
+        v.visit_query(query);
+
+        if let Some(e) = v.err {
+            return Err(e);
+        }
+
+        // If any `Ident` is used to qualify `name` and is an item, then `name`
+        // might be a column.
+        if v.items.iter().any(|i| v.qualifiers.contains_key(i)) {
+            return Err(ambiguous_err(name, "column"));
+        }
+
+        // If we see `<qual>.<item>`, but never see `<other>.<qual>.<item>`,
+        // then `<item>` is used as a schema.
+        if v.qualifiers.values().any(|v| v.is_empty()) {
+            return Err(ambiguous_err(name, "schema"));
+        }
```

Comment:
Er, what does it mean to use an item as a schema? Deleting this check doesn't cause any tests to fail.